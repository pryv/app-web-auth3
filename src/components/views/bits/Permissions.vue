<template>
  <v-dialog
    v-model="dialog"
    width="600"
    persistent>
    <v-card>
      <v-card-title
        id="appIdText"
        class="headline grey lighten-2">
        <span><b>{{ accessState.requestingAppId }}</b></span>
      </v-card-title>
      <v-card-text style="text-align:left">
        <h4>Is requesting permission:</h4>
        <ul>
          <li
            v-for="(permission, index) in accessState.requestedPermissions"
            :key="index">
            to {{ permission.level.toLowerCase() }} <u>{{ permission.streamId === '*' ? '* (all data)' : permission.name || permission.defaultName }}</u>
          </li>
        </ul>
        <span
          v-if="expire"
          style="text-align:left">
          <b>will expire after:</b> {{ accessState.expireAfter }}s
        </span>
        <br>
        <span
          v-if="consentMsg"
          v-html="consentMsg"/>
      </v-card-text>
      <v-alert
        v-if="accessMismatch"
        :value="accessMismatch"
        :type="alertType"
        transition="scale-transition"
        style="text-align:left">
        <b>A different access was already given to this app. Do you want to replace it?</b>
      </v-alert>
      <v-divider/>
      <v-card-actions>
        <v-spacer/>
        <v-btn
          id="refusePermissions"
          @click="closeDialog('refused')"
        >Reject</v-btn>
        <v-btn
          id="acceptPermissions"
          @click="closeDialog('accepted')"
        >Accept</v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script>
import { marked } from 'marked';
export default {
  props: {
    ctx: { type: Object, default: () => {} },
  },
  data: () => ({
    dialog: true,
    accessState: null,
    checkAppResult: null,
  }),

  computed: {
    consentMsg: function () {
      if (this.ctx.accessState.clientData != null) {
        const description = this.ctx.accessState.clientData['app-web-auth:description'];
        return description != null ? marked.parse(description.content) : '';
      } else {
        return null;
      }
    },
    accessMismatch: function () {
      return (this.ctx.checkAppResult.mismatchingAccess != null);
    },
    expire: function () {
      return (this.ctx.accessState.expireAfter != null);
    },
  },
  created: function () {
    this.accessState = this.ctx.accessState;
    this.checkAppResult = this.ctx.checkAppResult;
  },
  methods: {
    closeDialog (acceptOrReject) {
      this.dialog = false;
      this.$emit(acceptOrReject);
    },
  },
};
</script>
