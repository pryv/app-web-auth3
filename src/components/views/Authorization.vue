<template>
  <div>
    <h1>Sign in</h1>

    <Permissions
      v-if="checkedPermissions!=null"
      :appId="ctx.appId"
      :permissionsList="checkedPermissions"
      @accepted="accept"
      @refused="refuse"/>

    <v-form
      ref="form"
      v-model="validForm"
      @submit.prevent>

      <v-text-field
        id="usernameOrEmail"
        v-model="ctx.user.username"
        :rules="[rules.required]"
        :label="$t('usernameOrEmail')"/>

      <Password v-model="password"/>

      <v-btn
        id="submitButton"
        :disabled="!validForm||submitting"
        @click="submit"
      >{{ $t('signin') }}</v-btn>

      <v-btn
        @click="refuse"
      >{{ $t('cancel') }}</v-btn>

      <div v-if="serviceInfos.support">
        {{ $t('reachUs1') }}
        <a
          :href="serviceInfos.support"
          target="_blank">
          helpdesk</a>
        {{ $t('reachUs2') }}
      </div>
    </v-form>

    <v-divider class="mt-3 mb-2"/>

    <router-link :to="{ name: 'RegisterUser' }"><h3>{{ $t('register') }}</h3></router-link>

    <router-link :to="{ name: 'ResetPassword' }"><h3>{{ $t('forgotPass') }}</h3></router-link>

    <Alerts
      :errorMsg="error"/>
  </div>
</template>

<script>
import Password from './bits/Password.vue';
import Permissions from './bits/Permissions.vue';
import Alerts from './bits/Alerts.vue';
import Context from '../../context.js';
import controllerFactory from '../controller/controller.js';
import { i18n } from '../../locals/i18n.js';

export default {
  components: {
    Password,
    Permissions,
    Alerts,
  },
  data: () => ({
    password: '',
    personalToken: '',
    error: '',
    checkedPermissions: null,
    accessId: null,
    serviceInfos: {},
    submitting: false,
    c: null,
    ctx: {},
    rules: {
      required: value => !!value || i18n.t('requireField'),
      email: value => /.+@.+/.test(value) || i18n.t('invalidEmail'),
    },
    validForm: false,
  }),
  created () {
    this.ctx = new Context(this.$route.query);
    this.c = controllerFactory(this.ctx);
    this.c.getServiceInfo()
      .then(this.showInfos)
      .catch(this.showError);
  },
  methods: {
    submit () {
      if (this.$refs.form.validate()) {
        this.submitting = true;
        // Check for existing app access
        this.c.checkAccess(this.password, this.showPermissions)
          .catch(this.showError)
          .finally(() => { this.submitting = false; });
      }
    },
    // Print requested permissions to the user
    showPermissions (accessId) {
      this.checkedPermissions = this.ctx.permissions.list;
      this.accessId = accessId;
    },
    // The user accepts the requested permissions
    accept () {
      this.c.acceptAccess(this.accessId)
        .catch(this.showError);
    },
    // The user refuses the requested permissions
    refuse () {
      this.c.refuseAccess()
        .catch(this.showError);
    },
    showError (error) {
      this.error = error.msg;
    },
    showInfos (infos) {
      this.serviceInfos = infos;
    },
  },
};
</script>
