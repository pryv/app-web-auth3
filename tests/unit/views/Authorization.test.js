import { shallowMount } from '@vue/test-utils';
import Authorization from '@/components/views/Authorization';
import Permissions from '@/components/views/bits/Permissions';
import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.use(Vuetify);

describe('Authorization.test.js', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Authorization, {
      mocks: {
        $route: {query: {}},
        $t: (key) => key,
      },
      stubs: ['router-link'],
    });
  });

  it('renders correctly (snapshots matching)', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('deactivates submit button when form is not yet filled', () => {
    const submitButton = wrapper.find('#submitButton');
    expect(wrapper.vm.validForm).toBe(false);
    expect(submitButton.attributes().disabled).toBe('true');
  });

  it('activates submit button when form is valid', () => {
    wrapper.setData({
      validForm: true,
    });
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.attributes().disabled).toBeFalsy();
  });

  it('validates required fields correctly', () => {
    const requiredRule = wrapper.vm.rules.required;
    expect(requiredRule()).toBe('This field is required');
    expect(requiredRule('notEmpty')).toBe(true);
  });

  it('shows authorization form by default', () => {
    // Renders auth form
    const authForm = wrapper.find({ref: 'form'});
    expect(authForm.exists()).toBe(true);
    // Does not render permissions
    const permissions = wrapper.find(Permissions);
    expect(permissions.exists()).toBe(false);
  });

  it('shows requested permissions when app access check is successful', () => {
    wrapper.setData({
      checkedPermissions: [{streamId: 'diary', level: 'read'}],
    });
    // Renders permissions instead
    const permissions = wrapper.find(Permissions);
    expect(permissions.exists()).toBe(true);
  });
});
