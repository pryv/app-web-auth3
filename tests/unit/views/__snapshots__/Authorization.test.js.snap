// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization.test.js renders correctly (snapshots matching) 1`] = `
<div>
  <h1>
    Sign in
  </h1>
   
  <!---->
   
  <v-dialog-stub
    contentclass=""
    maxwidth="none"
    origin="center center"
    persistent="true"
    transition="dialog-transition"
    width="600"
  >
    <v-card-stub
      tag="div"
    >
      <v-card-title-stub
        class="headline grey lighten-2"
      >
        MFA verification
      </v-card-title-stub>
       
      <v-text-field-stub
        class="ma-3"
        id="mfaCode"
        label="MFA code"
        rules="function required(value) {
          return !!value || 'This field is required.';
        }"
        value=""
      />
       
      <v-card-actions-stub>
        <v-spacer-stub />
         
        <v-btn-stub
          activeclass="v-btn--active"
          tag="button"
          type="button"
        >
          Cancel
        </v-btn-stub>
         
        <v-btn-stub
          activeclass="v-btn--active"
          tag="button"
          type="button"
        >
          Ok
        </v-btn-stub>
      </v-card-actions-stub>
    </v-card-stub>
  </v-dialog-stub>
   
  <v-form-stub>
    <v-text-field-stub
      id="usernameOrEmail"
      label="Username or email"
      rules="function required(value) {
          return !!value || 'This field is required.';
        }"
      value=""
    />
     
    <password-stub
      value=""
    />
     
    <v-btn-stub
      activeclass="v-btn--active"
      disabled="true"
      id="submitButton"
      tag="button"
      type="button"
    >
      Sign In
    </v-btn-stub>
     
    <v-btn-stub
      activeclass="v-btn--active"
      tag="button"
      type="button"
    >
      Cancel
    </v-btn-stub>
     
    <!---->
  </v-form-stub>
   
  <v-divider-stub
    class="mt-3 mb-2"
  />
   
  <router-link-stub
    to="[object Object]"
  >
    <h3>
      Create an account
    </h3>
  </router-link-stub>
   
  <router-link-stub
    to="[object Object]"
  >
    <h3>
      Forgot password
    </h3>
  </router-link-stub>
   
  <alerts-stub
    errormsg=""
    successmsg=""
  />
</div>
`;
